src/plugins/discover/common/esql_locator.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { DataViewSpec } from '@kbn/data-views-plugin/common';
+import { DISCOVER_ESQL_LOCATOR } from '@kbn/deeplinks-analytics';
+import { LocatorDefinition, LocatorPublic } from '@kbn/share-plugin/common';
+import { SerializableRecord } from '@kbn/utility-types';
+
+export type DiscoverEsqlLocatorParams = SerializableRecord;
+
+export interface DiscoverEsqlLocatorDependencies {
+  discoverAppLocator: LocatorPublic<SerializableRecord>;
+  getIndices: (options: {
+    showAllIndices: boolean;
+    pattern: string;
+    isRollupIndex: () => boolean;
+  }) => Promise<Array<{ name: string }>>;
+  createDataViewSpec: (title: string) => Promise<DataViewSpec>;
+}
+
+export type DiscoverEsqlLocator = LocatorPublic<DiscoverEsqlLocatorParams>;
+
+export class DiscoverEsqlLocatorDefinition implements LocatorDefinition<DiscoverEsqlLocatorParams> {
+  public readonly id = DISCOVER_ESQL_LOCATOR;
+
+  constructor(protected readonly deps: DiscoverEsqlLocatorDependencies) {}
+
+  public readonly getLocation = async () => {
+    const { discoverAppLocator, getIndices } = this.deps;
+
+    const getIndicesList = async () => {
```

Comment:
Plan to move this logic to utils when this is merged https://github.com/elastic/kibana/pull/174736/